### PR TITLE
Sparse benchmark 

### PR DESCRIPTION
**Pull request by kalyc**
### Summary
Added primary benchmarking model for sparse with synthetic data

### Related Issues
* [Not supported operators in Keras-MXNet backend](https://github.com/awslabs/keras-apache-mxnet/issues/18)
* [Not supported examples in Keras-MXNet backend](https://github.com/awslabs/keras-apache-mxnet/issues/79)

### PR Overview

- [n] This PR requires new unit tests [y/n] (make sure tests are included)
- [y] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n]

### CODE REVIEW

File: benchmark/sparse/linear_regression/keras_tf_model.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import numpy as np
+from keras.layers import Dense, Input
+from keras.models import Sequential
+from keras.optimizers import SGD
+from scipy.sparse import csc_matrix
+
+# Fix random seed for reproducibility
+np.random.seed(7)
+
+start = time.time()
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_tf_model.py]: It's better to only time the training time. As data generation code is different in pure mxnet (extra step to crate Iterators)

File: benchmark/sparse/linear_regression/keras_tf_model.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import time
+
+import numpy as np
+from keras.layers import Dense, Input
+from keras.models import Sequential
+from keras.optimizers import SGD
+from scipy.sparse import csc_matrix
+
+# Fix random seed for reproducibility
+np.random.seed(7)
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_tf_model.py]: Can we create a common file to run the benchmark and it will create a random dataset instead of fixed seed? Some thing like `run_sparse_benchmark.py`, and the logic will be:
```
1. Generate random dataset
2. Pass the dataset to `keras_tf_model.py` to train and it returns acc, loss, time to train
3. Pass the dataset to `mxnet_sparse_model.py` to train and it returns acc, loss, time to train
```

In this way you can make sure same data is passed and each time you run it, data will be random generated

File: benchmark/sparse/linear_regression/keras_tf_model.py (hunk 1 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/keras_tf_model.py]: Good point, fixed.

File: benchmark/sparse/linear_regression/keras_tf_model.py (hunk 2 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/keras_tf_model.py]: Done

File: benchmark/sparse/linear_regression/keras_tf_model.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    predictions = Dense(10, activation='linear')(inputs)
+    model = Model(input=inputs, output=predictions)
+
+    sgd = SGD(lr=0.1, momentum=0.9)
+
+    model.compile(loss='mse', optimizer=sgd, metrics=['accuracy'])
+
+    start = time.time()
+    model.fit(train_data, train_label,
+              epochs=1000,
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_tf_model.py]: epoch can be a configurable param

File: benchmark/sparse/linear_regression/run_sparse_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import keras_tf_model
+import mxnet_sparse_model
+import numpy as np
+from scipy.sparse import csc_matrix
+
+
+def invoke_benchmark(batch_size, mode):
+    # Fix random seed for reproducibility
+    np.random.seed(7)
```
Comment (roywei) [inline benchmark/sparse/linear_regression/run_sparse_benchmark.py]: Now we can remove this fixed seed since same data is passing to both modes.

File: benchmark/sparse/linear_regression/results.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+### Configuration
+| Dataset          | Synthetic(Randomly generated)                                |
+| :--------------- | :----------------------------------------------------------- |
+| Keras            | v2.2.0                                                       |
+| TensorFlow       | v1.10.0                                                      |
+| MXNet            | v1.2.1                                                       |
+
+
+### Results 
+| Instance Type | GPUs  | Batch Size  | MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
```
Comment (roywei) [inline benchmark/sparse/linear_regression/results.md]: We can mention here if users want to reproduce exact result, they can use fixed seed 7.  In general, we should observe consist/similar result with different random datasets
Or we can show an average of a few runs.

Comment (kalyc) [reply benchmark/sparse/linear_regression/results.md]: Fixed

File: benchmark/sparse/linear_regression/run_sparse_benchmark.py (hunk 4 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/run_sparse_benchmark.py]: Done

File: benchmark/sparse/linear_regression/keras_tf_model.py (hunk 3 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/keras_tf_model.py]: Done

File: .travis.yml [hunk 6]
```diff
@@ -124,4 +124,4 @@ script:
       PYTHONPATH=$PWD:$PYTHONPATH py.test tests/test_documentation.py;
     else
       PYTHONPATH=$PWD:$PYTHONPATH py.test tests/ --ignore=tests/integration_tests --ignore=tests/test_documentation.py --ignore=tests/keras/legacy/layers_test.py --cov-config .coveragerc --cov=keras tests/;
-    fi
+    fi
```
Comment (roywei) [inline .travis.yml]: keep the same with dev branch

Comment (kalyc) [reply .travis.yml]: It is the same as dev - for some reason its showing that there is a diff.

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    print("MXNet Sparse Benchmark Results")
+    print("Batch Size")
+    print(batch_size)
+    print('Total Time')
+    print(time.time() - start)
+
+    metric = mx.metric.MSE()
+    mse = model.score(eval_iter, metric)
+    print("Achieved {0:.6f} validation MSE".format(mse[0][1]))
+    assert model.score(eval_iter, metric)[0][1] < 0.01001, "Achieved MSE (%f) is larger than expected (0.01001)" % \
```
Comment (sandeep-krishnamurthy) [inline benchmark/sparse/linear_regression/mxnet_sparse_model.py]: nit: why assert in benchmark script? 
Same comment for above tf script.

File: benchmark/sparse/linear_regression/results.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+## Summary
+```
+   Results below show the performance comparison of linear regression with MXNet vs Keras-Tensorflow using sparse tensors
+```
+
+### Configuration
+| Dataset          | Synthetic(Randomly generated)                                |
+| :--------------- | :----------------------------------------------------------- |
+| Keras            | v2.2.0                                                       |
```
Comment (sandeep-krishnamurthy) [inline benchmark/sparse/linear_regression/results.md]: 2.2.2?

File: benchmark/sparse/linear_regression/results.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+### Results 
+| Instance Type | GPUs  | Batch Size  | MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
+|-----|-----|-----|----- |-----|
+|  C5.18X Large |   0  | 16  | 51.15 sec | 71.85 sec |
+|  C5.18X Large |   0  | 32 | 20.82 sec | 54.31 sec |
+|  C5.18X Large |   0  | 64  | 13.13 sec | 19.20 sec |
+|  C5.18X Large |   0  | 128  | 5.72 sec | 9.86 sec |
+
+### Note
+For reproducing above results set seed to `7` by adding this line in the `run_sparse_benchmark` script - `np.random.seed(7)`
```
Comment (sandeep-krishnamurthy) [inline benchmark/sparse/linear_regression/results.md]: why don't we add this line rather than asking the user to add.

File: benchmark/sparse/linear_regression/results.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+Run the file as `python run_sparse_benchmark.py --batch=128 --epochs=1000`
```
Comment (sandeep-krishnamurthy) [inline benchmark/sparse/linear_regression/results.md]: why 1000 epochs? 25 should be good enough?

File: benchmark/sparse/linear_regression/results.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Linear Regression Benchmark Results 
+
+## Summary
+```
+   Results below show the performance comparison of linear regression with MXNet vs Keras-Tensorflow using sparse tensors
+```
+
+### Configuration
```
Comment (sandeep-krishnamurthy) [inline benchmark/sparse/linear_regression/results.md]: Please add a comment about sparse type - csr

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py (hunk 7 above)
Comment (sandeep-krishnamurthy) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: Also, we are generating random data, so accuracy/loss doesn't add any value here?

File: benchmark/sparse/linear_regression/run_sparse_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+import argparse
+
+import keras_tf_model
+import mxnet as mx
+import mxnet_sparse_model
+from scipy import sparse
+
+
+def invoke_benchmark(batch_size, epochs):
```
Comment (sandeep-krishnamurthy) [inline benchmark/sparse/linear_regression/run_sparse_benchmark.py]: Nit: it would be useful going forward if you have 2 function invoke_tf_keras_benchmark() and invoke_mxnet_benchmark(). and from CLI take an option for framework name may be?

File: benchmark/sparse/linear_regression/results.md (hunk 9 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/results.md]: We wanted to avoid the assumption that this benchmark is reproducible only one one dataset. So as per comments by @roywei we moved it under ReadMe - for details about how to get these values.

File: benchmark/sparse/linear_regression/run_sparse_benchmark.py (hunk 12 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/run_sparse_benchmark.py]: Yes that was in the earlier revisions - we decided to just go forward with having only one function with invoking both the benchmarks.

File: benchmark/sparse/linear_regression/results.md (hunk 10 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/results.md]: 25 doesn't converge well for either of the backends - when we have a MSE limit of `0.01001` set
MXNet model results in MSE `0.020521`
Keras-TF results in MSE `0.017107` with 25 epochs

File: benchmark/sparse/linear_regression/results.md (hunk 8 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/results.md]: updated

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py (hunk 7 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: For first comment - I followed the style of the benchmarking script for linear regression model written in MXNet - https://mxnet.incubator.apache.org/tutorials/python/linear-regression.html

As we have a pre-defined test and train set, even if its synthetic we can loss/accuracy can still help us with understanding how the model is performing.

Comment (kalyc) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: Didn't track memory usage as we just wanted to check speed of mxnet vs keras-tf on sparse tensors.

Comment (roywei) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: Agree that accuracy/loss has no value for synthetic data, even if the acc increase, it does not mean it's performing better. 
You can extend that in future if you want to include real dataset. and only check acc/loss on real dataset.

File: benchmark/sparse/linear_regression/results.md (hunk 9 above)
Comment (roywei) [reply benchmark/sparse/linear_regression/results.md]: Or we can report an average of multiple runs. My point is better performance on a fixed seed is not convinced enough.

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py (hunk 7 above)
Comment (roywei) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: About memory consumption , it's a good value add as sparse tensors save a lot of memory, and it's very easy to report. Just run on a GPU instance and report the output of `nvidia-smi`

Comment (kalyc) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: Yes thats a good point - I will update the memory usage in another PR

File: .travis.yml (hunk 6 above)
Comment (sandeep-krishnamurthy) [reply .travis.yml]: Diff is an extra line at end of file is deleted.
